### PR TITLE
Minio native support #2040

### PR DESCRIPTION
**Pull request by JiriOndrusek**
fixes https://github.com/apache/camel-quarkus/issues/2040

supersedes https://github.com/apache/camel-quarkus/pull/2102

Extension leverages quarkiverse-minio. There are several problems/limitations:

- It seems that there is a bug in quarkiverse-minio - it misses some reflection registrations (which are probably not needed in simpler scenario) - I'll try to report/fix it in quarkiverse-minio extension and then remove registration from camel-quarkus extension.
- There is a limitation because of quiarkiverse-minio which prevents to use more different minio endpoint at once. (minio client has to be defined in application.properties). Limitation is described in doc. I see an option in update of camel-minio component to allow overriding of autowired client by endpoint's configuration. @jamesnetherton , @ppalaga WDYT? Should I prepare an upgrade in camel-minio component?

 ] An issue should be filed for the change unless this is a trivial change (fixing a typo or similar). One issue should ideally be fixed by not more than one commit and the other way round, each commit should fix just one issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful and properly spelled subject line and body. Copying the title of the associated issue is typically enough. Please include the issue number in the commit message prefixed by #.
[ ] The pull request description should explain what the pull request does, how, and why. If the info is available in the associated issue or some other external document, a link is enough.
[ ] Phrases like Fix #<issueNumber> or Fixes #<issueNumber> will auto-close the named issue upon merging the pull request. Using them is typically a good idea.
[ ] Please run mvn process-resources -Pformat (and amend the changes if necessary) before sending the pull request.
[ ] Contributor guide is your good friend: https://camel.apache.org/camel-quarkus/latest/contributor-guide.html

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: docs/modules/ROOT/pages/reference/extensions/minio.adoc [hunk 1]
```diff
@@ -31,3 +31,11 @@ Please refer to the above link for usage and configuration details.
 ----
 
 Check the xref:user-guide/index.adoc[User guide] for more information about writing Camel Quarkus applications.
+
+== Camel Quarkus limitations
+
+This extension leverages the Quarkiverse Minio. Standard application.properties mechanism is used to define connection (see http://github.com/quarkiverse/quarkiverse-minio#configuration-reference[documentation]).
+
+It is mandatory to configure the Minio client in properties and Camel will autowire client into the Mino component.
```
Comment (aldettinger) [inline docs/modules/ROOT/pages/reference/extensions/minio.adoc:39]: Small typo here Mino => Minio ?

**Review by aldettinger — APPROVED** [review]
Well done, the MinioConfigProvider is interesting.
+1 to remove extra reflections once fixed upstream.
I have no strong opinion concerning the modification of the camel-minio component.

File: integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTestResource.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.camel.quarkus.testcontainers.ContainerResourceLifecycleManager;
+import org.apache.camel.util.CollectionHelper;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.HttpWaitStrategy;
+
+public class MinioTestResource implements ContainerResourceLifecycleManager {
+
+    public static final String CONTAINER_ACCESS_KEY = "MINIO_ACCESS_KEY";
+    public static final String CONTAINER_SECRET_KEY = "MINIO_SECRET_KEY";
+    private final String CONTAINER_IMAGE = "minio/minio:latest";
```
Comment (jamesnetherton) [inline integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTestResource.java:31]: Can we avoid the `latest` tag and pin to something else?

Comment (JiriOndrusek) [reply integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTestResource.java]: I've replaced latest with the current tag.

File: docs/modules/ROOT/pages/reference/extensions/minio.adoc (hunk 1 above)
Comment (JiriOndrusek) [reply docs/modules/ROOT/pages/reference/extensions/minio.adoc]: fixed

**Comment by JiriOndrusek** [comment]
Failed build is not linked to minio extension.

**Comment by ppalaga** [comment]
> There is a limitation because of quiarkiverse-minio which prevents to use more different minio endpoint at once. (minio client has to be defined in application.properties). Limitation is described in doc. I see an option in update of camel-minio component to allow overriding of autowired client by endpoint's configuration.

Sounds like a potentially important limitation. I think we should at least file a feature request on Quarkiverse, link the issue in our docs and advise users to upvote it if they need it. In that way we can see whether it is worth to invest our time into fixing it.

File: integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTest.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    @Test
+    public void testConsumer() throws Exception {
+        initClient();
+
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+
+        Future<String> fr1 = executor.submit(
+                () -> RestAssured.get("/minio/consumer")
+                        .then()
+                        .extract().asString());
```
Comment (ppalaga) [inline integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTest.java:57]: Do we really need the Executor? `consumerTemplate.receiveBody(String endpointUri, long timeout, Class<T> type)` is blocking so I guess you can put `RestAssured.get("/minio/consumer")` directly to the `untilAsserted()` block below.

Comment (JiriOndrusek) [reply integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTest.java:57]: thanks for the advice, I'll try it that way.

Comment (JiriOndrusek) [reply integration-tests/minio/src/test/java/org/apache/camel/quarkus/component/minio/it/MinioTest.java]: done

**Comment by JiriOndrusek** [comment]
@ppalaga I've created an [issue](https://github.com/quarkiverse/quarkiverse-minio/issues/22) about optional configuration of minio client in quarkiverse-minio and added to the doc how to upvote the issue if needed.

I've also created https://github.com/quarkiverse/quarkiverse-minio/issues/21 about missing registration for reflection and then I've created https://github.com/apache/camel-quarkus/issues/2120 to observe the issue.

**Review by ppalaga — APPROVED** [review]

**Comment by jamesnetherton** [comment]
@JiriOndrusek can you fix up conflicts and we can merge.

**Comment by JiriOndrusek** [comment]
I'll fix it

**Comment by JiriOndrusek** [comment]
@jamesnetherton  done

**Comment by jtama-op** [comment]
Fixes released in 0.2.0.

Warning, as per convention artifact will be prefixed with `quarkiverse-` from 0.3.0.

**Comment by JiriOndrusek** [comment]
Issue https://github.com/quarkiverse/quarkiverse-minio/issues/21 is fixed and incorporated into this PR

Unfortunately issue https://github.com/quarkiverse/quarkiverse-minio/issues/22 still causes problems, so limitation with configuration is still valid.

**Pull request merged**